spec/abci++/abci++_tmint_expected_behavior_002_draft.md

```diff
@@ -0,0 +1,209 @@
+
+non-proposer        = *got-vote [extend-non-proposer] *got-vote
+extend-non-proposer = process-proposal *got-vote [extend-vote]
+
+init-chain          = %s"<InitChain>"
+offer-snapshot      = %s"<OfferSnapshot>"
+apply-chunk         = %s"<ApplySnapshotChunk>"
+info                = %s"<Info>"
+prepare-proposal    = %s"<PrepareProposal>"
+process-proposal    = %s"<ProcessProposal>"
+extend-vote         = %s"<ExtendVote>"
+got-vote            = %s"<VerifyVoteExtension>"
+decide              = %s"<FinalizeBlock>"
+```
+
+>**TODO** We need to get the grammar reviewed by the people that know block-execution inside out.
+
+>**TODO** Still hesitating... introduce _n_ as total number of validators, so that we can bound the occurrences of
+>`got-vote` in a round.
+
+We have kept some of the ABCI++ methods out of the grammar, in order to keep it as clear and concise as possible.
+A common reason for keeping all these methods out is that they all can be called at any point in a sequence defined
+by the grammar above. Other reasons depend on the method in question:
+
+* `Echo` and `Flush` are only used for debugging purposes. Further, their handling by the Application should be trivial.
+* `CheckTx` is detached from the main method call sequence that drives block execution.
+* `Query` provides read-only access to the current Application state, so handling it should also be independent from
+  block execution.
+* Similarly, `ListSnapshots` and `LoadSnapshotChunk` provide read-only access to the Application's previously created
+  snapshots (if any), and help populate the parameters of `OfferSnapshot` and `ApplySnapshotChunk` at a process performing
+  state-sync while bootstrapping. Unlike `ListSnapshots` and `LoadSnapshotChunk`, both `OfferSnapshot`
+  and `ApplySnapshotChunk` _are_ included in the grammar.
+
+Finally, method `Info` is a special case. The method's purpose is three-fold, it can be used
+
+1. as part of handling an RPC call from an external client,
+2. as a handshake between Tendermint and the Application upon recovery to check whether any blocks need
+   to be replayed, and
+3. at the end of _state-sync_ to verify that the correct state has been reached.
+
+We have left `Info`'s first purpose out of the grammar for the same reasons as all the others: it can happen
+at any time, and has nothing to do with the block execution sequence. The second and third purposes, on the other
+hand, are present in the grammar.
+
+Let us now examine the grammar line by line, providing further details.
+
+* When a process starts, it may do so for the first time or after a crash (it is recovering).
+>```abnf
+>start               = clean-start / recovery
+>```
+
+* If the process is starting from scratch, Tendermint first calls `InitChain`, then it may optionally
+  start a _state-sync_ mechanism to catch up with other processes. Finally, it enters normal
+  consensus execution, which is a sequence of zero or more consensus heights.
```

Comment:
Shouldn't that be "one or more" consensus heights?

Reply:
That's a good point.
As it is now, the chain is also trying to capture beign processes that crash:
* Correct processes have an infinite output from the grammar
* Benign crashing processes (so, not correct) have a finite output, which can stop even before the first normally executed consensus height
Let's discuss this in person... I don't see a clear way to tackle this

Reply:
OK, I just pushed a commit tackling this. We have now an omega-grammar